File: companion-apps/web/src/integrations/hsl/HslRoutingAdapter.ts
import {
  type ActiveRouteSession,
  type CompanionSettings,
  type CoordinatePoint,
  CURRENT_ROUTE_PACKAGE_VERSION,
  type NormalizedRoutePackage,
  type RerouteContext,
  type RouteAlternative,
  type RouteManeuver,
  type RouteManeuverType,
  type RoutePlanRequest,
  type RoutePreviewModel,
  type RouteProviderID,
} from "../../domain/models.js";
import type { RoutingProvider } from "../../domain/providers.js";
import {
  approximateDistanceMeters,
  classifyTurn,
  cumulativeDistances,
  decodePolyline,
  deduplicateConsecutive,
  turnDeltaDegrees,
} from "../geo.js";
import { newAlternativeId, normalizedFromPreview } from "../routePackage.js";

export type HslSettingsProvider = () => CompanionSettings;

const ROUTE_PLAN_QUERY = `query RoutePlan($from: InputCoordinates!, $to: InputCoordinates!, $numItineraries: Int!, $transportModes: [TransportMode!]!, $optimize: OptimizeType!) {
  plan(from: $from, to: $to, numItineraries: $numItineraries, transportModes: $transportModes, optimize: $optimize) {
    itineraries { duration legs { mode distance from { lat lon name } to { lat lon name } legGeometry { points } } }
  }
}`;

type LivePlace = { lat: number; lon: number; name?: string };
type LiveLeg = {
  mode?: string;
  distance: number;
  from?: LivePlace;
  to?: LivePlace;
  legGeometry?: { points: string };
};
type LiveItinerary = { duration: number; legs: LiveLeg[] };
type DigitransitApiResponse = {
  data?: { plan?: { itineraries: LiveItinerary[] } };
  errors?: { message: string }[];
};

type Itinerary = {
  durationSeconds: number;
  systemNotice: string;
  legs: { distanceMeters: number; geometry: CoordinatePoint[] }[];
  startLabel: string;
  destinationLabel: string;
};

export class HslRoutingAdapter implements RoutingProvider {
  readonly providerID: RouteProviderID = "hsl";

  constructor(private readonly settingsProvider: HslSettingsProvider) {}

  async planRoute(request: RoutePlanRequest, signal?: AbortSignal): Promise<RoutePreviewModel> {
    return this.planPreview(request, undefined, signal);
  }

  async replanRoute(
    session: ActiveRouteSession,
    riderLocation: CoordinatePoint,
    rerouteContext?: RerouteContext,
    signal?: AbortSignal,
  ): Promise<RoutePreviewModel> {
    const rerouteOrigin = headingBiasedOrigin(riderLocation, rerouteContext, "hsl");
    const request: RoutePlanRequest = {
      origin: rerouteOrigin,
      destination: session.destinationCoordinate ?? riderLocation,
      providerID: session.providerID,
    };
    const revision = session.routeRevision === undefined ? undefined : session.routeRevision + 1;
    return this.planPreview(request, revision, signal);
  }

  normalizePreview(preview: RoutePreviewModel, request: RoutePlanRequest): NormalizedRoutePackage {
    return normalizedFromPreview(preview, request);
  }

  private async planPreview(
    request: RoutePlanRequest,
    revisionOverride: number | undefined,
    signal: AbortSignal | undefined,
  ): Promise<RoutePreviewModel> {
    const settings = this.settingsProvider();
    const cyclingSpeedKph = settings.cyclingSpeedKph;
    if (settings.preferLiveHslRouting) {
      const trimmedKey = settings.hslSubscriptionKey.trim();
      if (!trimmedKey) {
        return normalizeItineraries(
          sampleItineraries(request, "Fallback sample: missing HSL subscription key"),
          request,
          revisionOverride,
          "No HSL subscription key configured. Showing sample route instead.",
          cyclingSpeedKph,
        );
      }
      try {
        const live = await fetchLive(request, settings, signal);
        return normalizeItineraries(
          live,
          request,
          revisionOverride,
          "Live HSL Digitransit",
          cyclingSpeedKph,
        );
      } catch (err) {
        if ((err as Error)?.name === "AbortError") throw err;
        const message = err instanceof Error ? err.message : "Unknown error";
        return normalizeItineraries(
          sampleItineraries(request, "Fallback sample after live HSL failure"),
          request,
          revisionOverride,
          `Live HSL failed: ${message}. Showing sample route instead.`,
          cyclingSpeedKph,
        );
      }
    }
    return normalizeItineraries(
      sampleItineraries(request, "Sample HSL route"),
      request,
      revisionOverride,
      "Using sample HSL routes. Enable live HSL in Settings.",
      cyclingSpeedKph,
    );
  }
}

const MIN_HEADING_SPEED_MPS = 2.0;
const REROUTE_FORWARD_SHIFT_M = 15.0;

function headingBiasedOrigin(
  riderLocation: CoordinatePoint,
  rerouteContext: RerouteContext | undefined,
  providerLabel: string,
): CoordinatePoint {
  const heading = rerouteContext?.headingDegrees;
  const speedMps = rerouteContext?.speedMps;
  if (heading == null || !Number.isFinite(heading)) {
    console.debug(`[reroute_heading] provider=${providerLabel} reason=no_heading`);
    return riderLocation;
  }
  if (speedMps == null || !Number.isFinite(speedMps) || speedMps < MIN_HEADING_SPEED_MPS) {
    console.debug(`[reroute_heading] provider=${providerLabel} reason=low_speed speed=${speedMps ?? "nil"}`);
    return riderLocation;
  }
  const shifted = shiftPointByHeading(riderLocation, heading, REROUTE_FORWARD_SHIFT_M);
  if (
    !Number.isFinite(shifted.latitude) ||
    !Number.isFinite(shifted.longitude) ||
    (shifted.latitude === riderLocation.latitude && shifted.longitude === riderLocation.longitude)
  ) {
    console.debug(`[reroute_heading] provider=${providerLabel} reason=shift_failed`);
    return riderLocation;
  }
  console.debug(`[reroute_heading] provider=${providerLabel} reason=applied`);
  return shifted;
}

function shiftPointByHeading(
  point: CoordinatePoint,
  headingDegrees: number,
  distanceMeters: number,
): CoordinatePoint {
  const metersPerDegLat = 111_320.0;
  const rad = (headingDegrees * Math.PI) / 180;
  const northM = Math.cos(rad) * distanceMeters;
  const eastM = Math.sin(rad) * distanceMeters;
  const lat = point.latitude + northM / metersPerDegLat;
  const latRad = (point.latitude * Math.PI) / 180;
  const lonScale = metersPerDegLat * Math.cos(latRad);
  const lon = lonScale === 0 ? point.longitude : point.longitude + eastM / lonScale;
  return { latitude: lat, longitude: lon };
}

async function fetchLive(
  request: RoutePlanRequest,
  settings: CompanionSettings,
  signal?: AbortSignal,
): Promise<Itinerary[]> {
  const body = {
    query: ROUTE_PLAN_QUERY,
    variables: {
      from: { lat: request.origin.latitude, lon: request.origin.longitude },
      to: { lat: request.destination.latitude, lon: request.destination.longitude },
      numItineraries: 3,
      transportModes: [{ mode: "BICYCLE" }],
      optimize: "SAFE",
    },
  };
  const response = await fetch(settings.hslEndpointURL, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "digitransit-subscription-key": settings.hslSubscriptionKey,
    },
    body: JSON.stringify(body),
    signal,
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}`);
  }
  const decoded = (await response.json()) as DigitransitApiResponse;
  if (decoded.errors && decoded.errors.length > 0) {
    throw new Error(decoded.errors.map((e) => e.message).join(" | "));
  }
  const itineraries = decoded.data?.plan?.itineraries ?? [];
  if (itineraries.length === 0) {
    throw new Error("No HSL route alternatives were returned.");
  }
  return itineraries.map((itinerary, index) => ({
    durationSeconds: Math.round(itinerary.duration),
    systemNotice:
      index === 0 ? "HSL Digitransit live / fastest" : "HSL Digitransit live / alternative",
    legs: itinerary.legs
      .map((leg) => mapLiveLeg(leg))
      .filter(
        (leg): leg is { distanceMeters: number; geometry: CoordinatePoint[] } => leg !== null,
      ),
    startLabel: itinerary.legs[0]?.from?.name ?? "Current location",
    destinationLabel: itinerary.legs[itinerary.legs.length - 1]?.to?.name ?? "Selected destination",
  }));
}

function mapLiveLeg(leg: LiveLeg): { distanceMeters: number; geometry: CoordinatePoint[] } | null {
  const polylineGeometry = decodePolyline(leg.legGeometry?.points ?? "");
  const fallback: CoordinatePoint[] = [];
  if (leg.from) fallback.push({ latitude: leg.from.lat, longitude: leg.from.lon });
  if (leg.to) {
    const point = { latitude: leg.to.lat, longitude: leg.to.lon };
    const last = fallback[fallback.length - 1];
    if (!last || last.latitude !== point.latitude || last.longitude !== point.longitude) {
      fallback.push(point);
    }
  }
  const geometry = polylineGeometry.length >= 2 ? polylineGeometry : fallback;
  if (geometry.length < 2) return null;
  return { distanceMeters: leg.distance, geometry };
}

function normalizeItineraries(
  itineraries: Itinerary[],
  request: RoutePlanRequest,
  revisionOverride: number | undefined,
  planningNotice: string | undefined,
  cyclingSpeedKph: number,
): RoutePreviewModel {
  const alternatives = itineraries.map((itinerary, index) =>
    normalizeItinerary(itinerary, request, index, revisionOverride ?? 1, cyclingSpeedKph),
  );
  return {
    alternatives,
    selectedAlternativeID: alternatives[0]?.id,
    routeIdentifier: alternatives[0]?.normalizedPackage.routeIdentifier,
    routeRevision: alternatives[0]?.normalizedPackage.revision,
    planningNotice,
  };
}

function normalizeItinerary(
  itinerary: Itinerary,
  request: RoutePlanRequest,
  alternativeIndex: number,
  revision: number,
  cyclingSpeedKph: number,
): RouteAlternative {
  const geometry = deduplicatedGeometryFromLegs(itinerary.legs);
  const totalDistance = itinerary.legs.reduce((sum, leg) => sum + leg.distanceMeters, 0);
  const maneuvers = buildManeuvers(geometry, totalDistance);
  // Digitransit's bike speed is conservative for actual riders; recompute the
  // ETA from the user-set cycling speed so listed times match real-world riding.
  const durationSeconds = overrideDurationSeconds(
    totalDistance,
    cyclingSpeedKph,
    itinerary.durationSeconds,
  );
  const package_: NormalizedRoutePackage = {
    version: CURRENT_ROUTE_PACKAGE_VERSION,
    routeIdentifier: buildRouteIdentifier(request, alternativeIndex),
    revision,
    geometry,
    maneuvers,
    summary: {
      totalDistanceMeters: totalDistance,
      estimatedDurationSeconds: durationSeconds,
      startLabel: itinerary.startLabel,
      destinationLabel: itinerary.destinationLabel,
    },
    provenance: {
      providerID: "hsl",
      sourceReference: itinerary.systemNotice,
      generatedAtUnixMs: Date.now(),
    },
  };
  return {
    id: newAlternativeId(),
    title: alternativeIndex === 0 ? "Fastest bike route" : "Alternative bike route",
    subtitle: itinerary.systemNotice,
    distanceMeters: Math.round(totalDistance),
    durationSeconds,
    normalizedPackage: package_,
  };
}

function overrideDurationSeconds(
  totalDistanceMeters: number,
  cyclingSpeedKph: number,
  fallbackSeconds: number,
): number {
  if (!Number.isFinite(cyclingSpeedKph) || cyclingSpeedKph <= 0) return fallbackSeconds;
  const mps = cyclingSpeedKph / 3.6;
  return Math.max(1, Math.round(totalDistanceMeters / mps));
}

function deduplicatedGeometryFromLegs(legs: { geometry: CoordinatePoint[] }[]): CoordinatePoint[] {
  const points: CoordinatePoint[] = [];
  for (const leg of legs) {
    for (const point of leg.geometry) {
      const last = points[points.length - 1];
      if (!last || last.latitude !== point.latitude || last.longitude !== point.longitude) {
        points.push(point);
      }
    }
  }
  return points;
}

function buildManeuvers(geometry: CoordinatePoint[], totalDistance: number): RouteManeuver[] {
  const cumulative = cumulativeDistances(geometry);
  const maneuvers: RouteManeuver[] = [
    {
      id: "depart",
      maneuverType: "depart",
      location: geometry[0] ?? { latitude: 0, longitude: 0 },
      distanceFromStartMeters: 0,
      distanceToNextMeters: cumulative[1],
      instructionText: "Start riding",
    },
  ];
  for (let i = 1; i < geometry.length - 1; i++) {
    const delta = turnDeltaDegrees(geometry[i - 1], geometry[i], geometry[i + 1]);
    const turn = classifyTurn(delta);
    if (!turn) continue;
    const distanceToNext =
      i + 1 < cumulative.length ? cumulative[i + 1] - cumulative[i] : undefined;
    maneuvers.push({
      id: `step-${maneuvers.length}`,
      maneuverType: turn.type as RouteManeuverType,
      location: geometry[i],
      distanceFromStartMeters: cumulative[i],
      distanceToNextMeters: distanceToNext,
      instructionText: turn.instruction,
    });
  }
  maneuvers.push({
    id: "arrive",
    maneuverType: "arrive",
    location: geometry[geometry.length - 1] ?? { latitude: 0, longitude: 0 },
    distanceFromStartMeters: totalDistance,
    instructionText: "Arrive at destination",
  });
  return maneuvers;
}

function buildRouteIdentifier(request: RoutePlanRequest, alternativeIndex: number): string {
  const o = `${request.origin.latitude.toFixed(5)},${request.origin.longitude.toFixed(5)}`;
  const d = `${request.destination.latitude.toFixed(5)},${request.destination.longitude.toFixed(5)}`;
  return `hsl:${o}->${d}:alt-${alternativeIndex}`;
}

function sampleItineraries(request: RoutePlanRequest, liveDescriptor: string): Itinerary[] {
  const variants: { tag: string; offsetScale: number; pattern: number[] }[] = [
    {
      tag: "fastest",
      offsetScale: 0.0013,
      pattern: [0.26, 0.52, 0.22, 0.0, 0.16, 0.04],
    },
    {
      tag: "quieter",
      offsetScale: 0.0016,
      pattern: [-0.2, -0.42, -0.18, -0.28, -0.1, 0.02],
    },
    {
      tag: "simpler",
      offsetScale: 0.001,
      pattern: [0.12, 0.06, 0.28, 0.14, 0.24, 0.08],
    },
  ];
  return variants.map(({ tag, offsetScale, pattern }) => {
    const geometry = sampleGeometry(request.origin, request.destination, offsetScale, pattern);
    const distance = geometry
      .slice(1)
      .reduce((sum, p, i) => sum + approximateDistanceMeters(geometry[i], p), 0);
    return {
      durationSeconds: Math.round(distance / 4.2),
      systemNotice: `${liveDescriptor} / ${tag}`,
      legs: [{ distanceMeters: distance, geometry }],
      startLabel: "Current location",
      destinationLabel: "Selected destination",
    };
  });
}

function sampleGeometry(
  origin: CoordinatePoint,
  destination: CoordinatePoint,
  offsetScale: number,
  pattern: number[],
): CoordinatePoint[] {
  if (origin.latitude === destination.latitude && origin.longitude === destination.longitude) {
    return deduplicateConsecutive([
      origin,
      { latitude: origin.latitude + 0.0015, longitude: origin.longitude + 0.0009 },
      { latitude: origin.latitude + 0.0024, longitude: origin.longitude + 0.0016 },
      { latitude: origin.latitude + 0.0019, longitude: origin.longitude - 0.0004 },
      { latitude: origin.latitude + 0.0008, longitude: origin.longitude - 0.0016 },
      origin,
    ]);
  }
  const latDelta = destination.latitude - origin.latitude;
  const lonDelta = destination.longitude - origin.longitude;
  const length = Math.max(Math.sqrt(latDelta * latDelta + lonDelta * lonDelta), 0.0001);
  const perpendicularLat = -lonDelta / length;
  const perpendicularLon = latDelta / length;
  const fractions = [0.1, 0.22, 0.38, 0.54, 0.72, 0.88];
  const geometry: CoordinatePoint[] = [origin];
  fractions.forEach((fraction, index) => {
    const lateral = offsetScale * pattern[index];
    const forwardBias = offsetScale * 0.1 * (index - 2.5);
    geometry.push({
      latitude:
        origin.latitude + latDelta * fraction + perpendicularLat * lateral + latDelta * forwardBias,
      longitude:
        origin.longitude +
        lonDelta * fraction +
        perpendicularLon * lateral +
        lonDelta * forwardBias,
    });
  });
  geometry.push(destination);
  return geometry;
}

File: companion-apps/ios/CompanionApp/Integration/HSL/HslRoutingAdapter.swift
import Foundation

struct HslRoutingAdapter: RoutingProvider {
    let providerID: RouteProviderID = .hsl
    let isAvailableInV1: Bool = true

    private let settingsProvider: () -> CompanionSettings

    init(settingsProvider: @escaping () -> CompanionSettings = { .defaults }) {
        self.settingsProvider = settingsProvider
    }

    func planRoute(_ request: RoutePlanRequest) async throws -> RoutePreviewModel {
        try await planPreview(for: request, revisionOverride: nil)
    }

    func replanRoute(
        using session: ActiveRouteSession,
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?
    ) async throws -> RoutePreviewModel {
        let rerouteOrigin = headingBiasedOrigin(
            riderLocation: riderLocation,
            rerouteContext: rerouteContext,
            providerLabel: "hsl"
        )
        let rerouteRequest = RoutePlanRequest(
            origin: rerouteOrigin,
            destination: session.destinationCoordinate ?? riderLocation,
            providerID: session.providerID
        )
        let revisionOverride = session.routeRevision.map { $0 + 1 }
        return try await planPreview(for: rerouteRequest, revisionOverride: revisionOverride)
    }

    private func headingBiasedOrigin(
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?,
        providerLabel: String
    ) -> CoordinatePoint {
        guard let heading = rerouteContext?.headingDegrees, heading.isFinite else {
            return riderLocation
        }
        guard let speed = rerouteContext?.speedMps, speed.isFinite, speed >= Self.minHeadingSpeedMps else {
            return riderLocation
        }
        let shifted = shiftPointByHeading(point: riderLocation, headingDegrees: heading, distanceMeters: Self.rerouteForwardShiftM)
        if !shifted.latitude.isFinite || !shifted.longitude.isFinite ||
            (shifted.latitude == riderLocation.latitude && shifted.longitude == riderLocation.longitude) {
            return riderLocation
        }
        return shifted
    }

    private func shiftPointByHeading(
        point: CoordinatePoint,
        headingDegrees: Double,
        distanceMeters: Double
    ) -> CoordinatePoint {
        let metersPerDegLat = 111_320.0
        let rad = headingDegrees * .pi / 180.0
        let northM = cos(rad) * distanceMeters
        let eastM = sin(rad) * distanceMeters
        let latitude = point.latitude + northM / metersPerDegLat
        let lonScale = metersPerDegLat * cos(point.latitude * .pi / 180.0)
        let longitude = lonScale == 0
            ? point.longitude
            : point.longitude + eastM / lonScale
        return CoordinatePoint(latitude: latitude, longitude: longitude)
    }

    func normalizePreview(_ preview: RoutePreviewModel, request: RoutePlanRequest) throws -> NormalizedRoutePackage {
        guard let selected = preview.selectedAlternative else {
            throw HslRoutingAdapterError.noAlternativesAvailable
        }
        return selected.normalizedPackage
    }

    private func planPreview(
        for request: RoutePlanRequest,
        revisionOverride: Int?
    ) async throws -> RoutePreviewModel {
        let settings = settingsProvider()
        let liveResponse = try await fetchLiveDigitransitResponse(for: request, settings: settings)
        return normalizeResponse(
            liveResponse,
            request: request,
            revisionOverride: revisionOverride,
            planningNotice: "Live HSL Digitransit"
        )
    }

    func makeGraphQLRequestBody(for request: RoutePlanRequest) -> DigitransitGraphQLRequestBody {
        DigitransitGraphQLRequestBody(
            query: Self.routePlanQuery,
            variables: .init(
                from: .init(lat: request.origin.latitude, lon: request.origin.longitude),
                to: .init(lat: request.destination.latitude, lon: request.destination.longitude),
                numItineraries: 3,
                transportModes: [.init(mode: "BICYCLE")],
                optimize: "SAFE"
            )
        )
    }

    private func fetchLiveDigitransitResponse(
        for request: RoutePlanRequest,
        settings: CompanionSettings
    ) async throws -> DigitransitResponse {
        guard let url = URL(string: settings.hslEndpointURL) else {
            throw HslRoutingAdapterError.invalidEndpoint(settings.hslEndpointURL)
        }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json", forHTTPHeaderField: "content-type")
        urlRequest.httpBody = try JSONEncoder().encode(makeGraphQLRequestBody(for: request))

        let (data, response) = try await URLSession.shared.data(for: urlRequest)
        guard let http = response as? HTTPURLResponse else {
            throw HslRoutingAdapterError.networkFailure("Missing HTTP response")
        }
        guard (200 ..< 300).contains(http.statusCode) else {
            let bodyMessage = String(data: data, encoding: .utf8) ?? "No response body"
            throw HslRoutingAdapterError.networkFailure("HTTP \(http.statusCode): \(bodyMessage)")
        }

        let decoded = try JSONDecoder().decode(DigitransitApiResponse.self, from: data)
        if let errors = decoded.errors, !errors.isEmpty {
            throw HslRoutingAdapterError.networkFailure(errors.map(\.message).joined(separator: " | "))
        }
        guard let plan = decoded.data?.plan, !plan.itineraries.isEmpty else {
            throw HslRoutingAdapterError.noAlternativesAvailable
        }

        let itineraries = plan.itineraries.enumerated().map { index, itinerary in
            let legs = itinerary.legs.compactMap { liveLeg in
                mapLiveLeg(liveLeg)
            }
            return DigitransitItinerary(
                durationSeconds: Int(itinerary.duration.rounded()),
                systemNotice: index == 0 ? "HSL Digitransit live / fastest" : "HSL Digitransit live / alternative",
                legs: legs,
                steps: [],
                startLabel: itinerary.legs.first?.from?.name ?? "Current location",
                destinationLabel: itinerary.legs.last?.to?.name ?? "Selected destination"
            )
        }

        return DigitransitResponse(
            data: DataContainer(
                plan: DigitransitPlan(itineraries: itineraries)
            )
        )
    }

    private func mapLiveLeg(_ leg: LiveLeg) -> DigitransitLeg? {
        let geometry = ShareImportUtilities.decodePolyline(leg.legGeometry?.points ?? "")
        let fallback = fallbackGeometry(from: leg)
        let points = geometry.count >= 2 ? geometry : fallback
        guard points.count >= 2 else { return nil }
        return DigitransitLeg(
            mode: leg.mode ?? "BICYCLE",
            distanceMeters: leg.distance,
            geometry: points
        )
    }

    private func fallbackGeometry(from leg: LiveLeg) -> [CoordinatePoint] {
        var points: [CoordinatePoint] = []
        if let from = leg.from {
            points.append(CoordinatePoint(latitude: from.lat, longitude: from.lon))
        }
        if let to = leg.to {
            let point = CoordinatePoint(latitude: to.lat, longitude: to.lon)
            if points.last != point {
                points.append(point)
            }
        }
        return points
    }

    private func normalizeResponse(
        _ response: DigitransitResponse,
        request: RoutePlanRequest,
        revisionOverride: Int?,
        planningNotice: String?
    ) -> RoutePreviewModel {
        let cyclingSpeedKph = settingsProvider().cyclingSpeedKph
        let alternatives = response.data.plan.itineraries.enumerated().map { index, itinerary in
            normalizeItinerary(
                itinerary,
                request: request,
                alternativeIndex: index,
                revision: revisionOverride ?? 1,
                cyclingSpeedKph: cyclingSpeedKph
            )
        }
        return RoutePreviewModel(
            alternatives: alternatives,
            selectedAlternativeID: alternatives.first?.id,
            routeIdentifier: alternatives.first?.normalizedPackage.routeIdentifier,
            routeRevision: alternatives.first?.normalizedPackage.revision,
            planningNotice: planningNotice
        )
    }

    private func normalizeItinerary(
        _ itinerary: DigitransitItinerary,
        request: RoutePlanRequest,
        alternativeIndex: Int,
        revision: Int,
        cyclingSpeedKph: Double
    ) -> RouteAlternative {
        let routeID = buildRouteIdentifier(request: request, alternativeIndex: alternativeIndex)
        let geometry = deduplicatedGeometry(from: itinerary.legs)
        let maneuvers = buildManeuvers(from: itinerary, geometry: geometry)
        let totalDistance = itinerary.legs.reduce(0.0) { $0 + $1.distanceMeters }
        // Digitransit's bike speed is conservative for actual riders; recompute
        // the ETA from the user-set cycling speed so listed times match
        // real-world riding.
        let durationSeconds = Self.overrideDurationSeconds(
            totalDistanceMeters: totalDistance,
            cyclingSpeedKph: cyclingSpeedKph,
            fallbackSeconds: itinerary.durationSeconds
        )
        let summary = RouteSummary(
            totalDistanceMeters: totalDistance,
            estimatedDurationSeconds: durationSeconds,
            startLabel: itinerary.startLabel,
            destinationLabel: itinerary.destinationLabel
        )
        let package = NormalizedRoutePackage(
            version: .current,
            routeIdentifier: routeID,
            revision: revision,
            geometry: geometry,
            maneuvers: maneuvers,
            summary: summary,
            provenance: RouteProvenance(
                providerID: .hsl,
                sourceReference: itinerary.systemNotice,
                generatedAtUnixMs: UInt64(Date().timeIntervalSince1970 * 1000)
            )
        )
        return RouteAlternative(
            id: UUID(),
            title: alternativeIndex == 0 ? "Fastest bike route" : "Alternative bike route",
            subtitle: itinerary.systemNotice,
            distanceMeters: Int(summary.totalDistanceMeters.rounded()),
            durationSeconds: summary.estimatedDurationSeconds,
            normalizedPackage: package
        )
    }

    static func overrideDurationSeconds(
        totalDistanceMeters: Double,
        cyclingSpeedKph: Double,
        fallbackSeconds: Int
    ) -> Int {
        guard cyclingSpeedKph.isFinite, cyclingSpeedKph > 0 else { return fallbackSeconds }
        let mps = cyclingSpeedKph / 3.6
        return max(1, Int((totalDistanceMeters / mps).rounded()))
    }

    private func buildManeuvers(
        from itinerary: DigitransitItinerary,
        geometry: [CoordinatePoint]
    ) -> [RouteManeuver] {
        let routeDistance = itinerary.legs.reduce(0.0) { $0 + $1.distanceMeters }
        let steps = itinerary.steps.isEmpty ? deriveSteps(from: geometry) : itinerary.steps
        var maneuvers: [RouteManeuver] = [
            RouteManeuver(
                id: "depart",
                maneuverType: .depart,
                location: geometry.first ?? CoordinatePoint(latitude: 0.0, longitude: 0.0),
                distanceFromStartMeters: 0.0,
                distanceToNextMeters: steps.first?.distanceFromStartMeters,
                instructionText: "Start riding"
            )
        ]

        for (index, step) in steps.enumerated() {
            maneuvers.append(
                RouteManeuver(
                    id: "step-\(index)",
                    maneuverType: maneuverType(for: step.relativeDirection),
                    location: step.location,
                    distanceFromStartMeters: step.distanceFromStartMeters,
                    distanceToNextMeters: step.distanceToNextMeters,
                    instructionText: step.instruction
                )
            )
        }

        maneuvers.append(
            RouteManeuver(
                id: "arrive",
                maneuverType: .arrive,
                location: geometry.last ?? CoordinatePoint(latitude: 0.0, longitude: 0.0),
                distanceFromStartMeters: routeDistance,
                distanceToNextMeters: nil,
                instructionText: "Arrive at destination"
            )
        )

        return maneuvers
    }

    private func deriveSteps(from geometry: [CoordinatePoint]) -> [DigitransitStep] {
        guard geometry.count >= 3 else { return [] }
        let cumulative = cumulativeDistances(geometry)
        var steps: [DigitransitStep] = []
        for index in 1 ..< geometry.count - 1 {
            let turnDelta = turnDeltaDegrees(previous: geometry[index - 1], current: geometry[index], next: geometry[index + 1])
            guard let classification = classifyTurn(deltaDegrees: turnDelta) else { continue }
            let distanceToNext = index + 1 < cumulative.count ? cumulative[index + 1] - cumulative[index] : nil
            steps.append(
                DigitransitStep(
                    relativeDirection: classification.token,
                    location: geometry[index],
                    distanceFromStartMeters: cumulative[index],
                    distanceToNextMeters: distanceToNext,
                    instruction: classification.instruction
                )
            )
        }
        return steps
    }

    private func classifyTurn(deltaDegrees: Double) -> (token: String, instruction: String)? {
        let magnitude = abs(deltaDegrees)
        guard magnitude >= 25 else { return nil }
        if magnitude >= 170 { return (deltaDegrees > 0 ? "UTURN_RIGHT" : "UTURN_LEFT", "Make a U-turn") }
        if magnitude >= 110 { return (deltaDegrees > 0 ? "HARD_RIGHT" : "HARD_LEFT", deltaDegrees > 0 ? "Turn sharply right" : "Turn sharply left") }
        if magnitude >= 50 { return (deltaDegrees > 0 ? "RIGHT" : "LEFT", deltaDegrees > 0 ? "Turn right" : "Turn left") }
        return (deltaDegrees > 0 ? "SLIGHTLY_RIGHT" : "SLIGHTLY_LEFT", deltaDegrees > 0 ? "Slight right" : "Slight left")
    }

    private func deduplicatedGeometry(from legs: [DigitransitLeg]) -> [CoordinatePoint] {
        var points: [CoordinatePoint] = []
        for point in legs.flatMap(\.geometry) {
            if points.last != point {
                points.append(point)
            }
        }
        return points
    }

    private func buildRouteIdentifier(request: RoutePlanRequest, alternativeIndex: Int) -> String {
        let origin = String(format: "%.5f,%.5f", request.origin.latitude, request.origin.longitude)
        let destination = String(format: "%.5f,%.5f", request.destination.latitude, request.destination.longitude)
        return "hsl:\(origin)->\(destination):alt-\(alternativeIndex)"
    }

    private func maneuverType(for relativeDirection: String) -> RouteManeuverType {
        switch relativeDirection.uppercased() {
        case "CONTINUE":
            return .straight
        case "SLIGHTLY_LEFT":
            return .slightLeft
        case "LEFT":
            return .left
        case "HARD_LEFT":
            return .sharpLeft
        case "SLIGHTLY_RIGHT":
            return .slightRight
        case "RIGHT":
            return .right
        case "HARD_RIGHT":
            return .sharpRight
        case "UTURN_LEFT", "UTURN_RIGHT", "UTURN":
            return .uturn
        case "CIRCLE_COUNTERCLOCKWISE", "CIRCLE_CLOCKWISE":
            return .roundabout
        case "ELEVATOR", "TRANSFER":
            return .ramp
        default:
            return .straight
        }
    }

    private static let minHeadingSpeedMps: Double = 2.0
    private static let rerouteForwardShiftM: Double = 15.0

    static let routePlanQuery = """
    query RoutePlan($from: InputCoordinates!, $to: InputCoordinates!, $numItineraries: Int!, $transportModes: [TransportMode!]!, $optimize: OptimizeType!) {
      plan(
        from: $from,
        to: $to,
        numItineraries: $numItineraries,
        transportModes: $transportModes,
        optimize: $optimize
      ) {
        itineraries {
          duration
          legs {
            mode
            distance
            from {
              lat
              lon
              name
            }
            to {
              lat
              lon
              name
            }
            legGeometry {
              points
            }
          }
        }
      }
    }
    """
}

enum HslRoutingAdapterError: LocalizedError {
    case noAlternativesAvailable
    case invalidEndpoint(String)
    case networkFailure(String)

    var errorDescription: String? {
        switch self {
        case .noAlternativesAvailable:
            return "No HSL route alternatives were returned."
        case .invalidEndpoint(let endpoint):
            return "Invalid HSL endpoint: \(endpoint)"
        case .networkFailure(let message):
            return message
        }
    }
}

